qiskit_ibm_provider/transpiler/passes/scheduling/dynamical_decoupling.py

```diff
@@ -154,85 +165,121 @@ def __init__(
-                    ]
-                    if gate_length % self._alignment != 0:
-                        # This is necessary to implement lightweight scheduling logic for this pass.
-                        # Usually the pulse alignment constraint and pulse data chunk size take
-                        # the same value, however, we can intentionally violate this pattern
-                        # at the gate level. For example, we can create a schedule consisting of
-                        # a pi-pulse of 32 dt followed by a post buffer, i.e. delay, of 4 dt
-                        # on the device with 16 dt constraint. Note that the pi-pulse length
-                        # is multiple of 16 dt but the gate length of 36 is not multiple of it.
-                        # Such pulse gate should be excluded.
-                        raise TranspilerError(
-                            f"Pulse gate {gate.name} with length non-multiple of {self._alignment} "
-                            f"is not acceptable in {self.__class__.__name__} pass."
-                        )
-                except KeyError:
-                    gate_length = self._durations.get(gate, physical_index)
-                sequence_lengths.append(gate_length)
-                # Update gate duration. This is necessary for current timeline drawer, i.e. scheduled.
-                gate.duration = gate_length
-            self._dd_sequence_lengths[qubit] = sequence_lengths
+            # Check if DD sequence is identity
+            if num_pulses != 1:
+                if num_pulses % 2 != 0:
+                    raise TranspilerError(
+                        "DD sequence must contain an even number of gates (or 1)."
+                    )
+                # TODO: this check should use the quantum info package in Qiskit.
+                noop = np.eye(2)
+                for gate in self._dd_sequences[seq_idx]:
+                    noop = noop.dot(gate.to_matrix())
+                if not matrix_equal(noop, IGate().to_matrix(), ignore_phase=True):
+                    raise TranspilerError(
+                        "The DD sequence does not make an identity operation."
+                    )
+                self._sequence_phase = np.angle(noop[0][0])
+
+            # Precompute qubit-wise DD sequence length for performance
+            for qubit in dag.qubits:
+                seq_length_ = []
+                if qubit not in self._dd_sequence_lengths:
+                    self._dd_sequence_lengths[qubit] = []
+
+                physical_index = dag.qubits.index(qubit)
+                if self._qubits and physical_index not in self._qubits:
+                    continue
+
+                for gate in seq:
+                    try:
+                        # Check calibration.
+                        gate_length = dag.calibrations[gate.name][
+                            (physical_index, gate.params)
+                        ]
```

Comment:
Likely this is a bug in terra but we should get `.duration` of the attached schedule.

Reply:
This is just a copy/paste of Qiskit's logic. Would appreciate a follow-up PR.